notes-0319-inductive-types.tex

```diff
@@ -0,0 +1,228 @@
+
+\begin{example}
+  $\2$ is generated by $0 : \2$ and $1 : \2$. Since neither constructor is a function type, there are no operations we can perform on these constructors. Therefore the type freely generated by these constructors has only two elements, as we would expect.
+\end{example}
+
+\begin{example}
+  $A + B$ is generated by $inl : A \to A + B$ and $inr : B \to A + B$. In this case the constructors are functions, so we get some more interesting elements of the inductive type. The type freely generated by these constructors contains elements like $inl\ a$ for elements $a : A$. 
+\end{example}
+    
+The first more interesting case of an inductive type is $\N$, which is generated by $zero : \N$ and $suc: \N \to \N$. This illustrates the fact that the type being defined can appear in both the domain and the co-domain of a constructor. The $suc$ constructor represents the construction of $n+1$ from a construction of $n$. Intuitively, the type freely generated by these constructors should contain $zero$, $suc(zero)$, $suc(suc(zero))$, and so on, since these are exactly the things we can construct from these constructors using the operations we have available. We can formalize this intuition with the following rules for the type $\N$: 
+
+\begin{multicols}{2}
+  \noindent
+  \begin{equation*}
+    \begin{prooftree}
+      \infer0[$0$-Introduction]{\oftype{zero}{\N}}
+    \end{prooftree}
+  \end{equation*}
+  \begin{equation*}
+    \begin{prooftree}
+      \hypo{\oftype{M}{\N}}
+      \infer1[$suc$-Introduction]{\oftype{suc\ M}{\N}}
+    \end{prooftree}
+  \end{equation*}
+\end{multicols}
+
+
+Then we would like the elimination rule to model the classic induction principle for the natural numbers. Recall that traditionally the induction principle for the set of natural numbers states that in order to prove some proposition $P(n)$ for all $n \in \N$ it suffices to prove $P(0)$ and $P(n) \rightarrow P(n+1)$. We can model this rule with the following axiom for our inductive $\N$ type:
+
+\begin{prooftree*}
+  \hypo{x : \N : \rightarrow P : U}
+  \hypo{M_0 : P[zero/x]}
+  \hypo{ O : \N}
+  \infer[no rule]3{y : P[x'/x] \rightarrow M_{suc} : P[(suc x' / x]}
+  \infer1[$N$-Elimination]{rec_\N [x.P] (M_0 ; x.y.M_{succ} ; O) : P[O/x]}
+\end{prooftree*}
+
+With traditional induction on natural numbers, induction is a black box, we don't necessarily know what the proofs for each of the $P(n)$ we are proving look like. In type theory, we want to go one step farther now and say exactly what the proof of $P(n)$ look like for each $n$. This motivates two computation rules, that tell us which proof of $P(n)$ (or equivalently which element $p : P(n)$) is constructed by the induction principle for the different values of $O$:
```

Comment:
Good point. Please cite the source: https://link.springer.com/chapter/10.1007/3-540-52335-9_47. Note that the result may or may not apply to Agda because the construction is using a special impredicative universe Prop, which is available in Coq but not in Agda.